devmapper: Emit warning messages when using loopback, update docs

cgwalters opened the pull request:
We really want people to use `dm.thinpooldev` as it's most performant.
Loopback is convenient for dev scenarios but the message "don't use
this in production" is not coming across strongly enough.  Fix that by
both noting it in the `README.md` here as well as by emitting warning
messages.

cgwalters committed:
devmapper: Emit warning messages when using loopback, update docs

We really want people to use `dm.thinpooldev` as it's most performant.
Loopback is convenient for dev scenarios but the message "don't use
this in production" is not coming across strongly enough.  Fix that by
both noting it in the `README.md` here as well as by emitting warning
messages.

Docker-DCO-1.1-Signed-off-by: Colin Walters <walters@verbum.org> (github: cgwalters)

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "devmapper-note-users-avoid-loopback" git@github.com:cgwalters/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

cgwalters commented:
Rebased and made `gofmt` happy; sorry this lost the earlier comments in https://github.com/cgwalters/docker/commit/d1e587bc8562e37c8e946191c1020a5f5504d766#commitcomment-10741168

icecrime commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1096,6 +1096,14 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 		}
 	}
 
+	if devices.thinPoolDevice == "" {
```
This is right below the closing brace of `if info.Exists == 0 && devices.thinPoolDevice == "" {`, there's probably some way to make this more readable.

icecrime commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1096,6 +1096,14 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 
+	if devices.thinPoolDevice == "" {
+		if devices.metadataLoopFile != "" || devices.dataLoopFile != "" {
+			logrus.Errorf("WARNING: LOOP DEVICE FALLBACK: No --storage-opt dm.thinpooldev specified; Usage of loopback is very strongly discouraged for production use.  Use --storage-opt dm.thinpooldev instead.  For more information see `man docker`")
```
I'm not sure `man docker` has much to say about this?

vbatts replied:
@icecrime  ah, we completed man pages for docker in fedora.

LK4D4 replied:
`logrus.Errorf("WARNING...)` looks little weird. It is error or warning, also you don't need `WARNING` word here.

icecrime commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1096,6 +1096,14 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
+		if devices.metadataLoopFile != "" || devices.dataLoopFile != "" {
+			logrus.Errorf("WARNING: LOOP DEVICE FALLBACK: No --storage-opt dm.thinpooldev specified; Usage of loopback is very strongly discouraged for production use.  Use --storage-opt dm.thinpooldev instead.  For more information see `man docker`")
+		} else {
+			logrus.Warnf("WARNING: Usage of --storage-opt dm.datadev or dm.metadatadev is strongly discouraged for production use.  Use --storage-opt dm.thinpooldev instead.  For more information see `man docker`")
```
I'm not sure `man docker` has much to say about this?

LK4D4 replied:
Same here.

LK4D4 commented:
@vbatts Can we have man pages from fedora?

cgwalters commented:
I addressed the documentation issues in https://github.com/docker/docker/pull/13323 - let's merge that first, and I'll get back to addressing the review comments for the code here after.

cgwalters commented:
Looks like this is obsoleted by https://github.com/docker/docker/pull/13266
